# Neural SINDy: Sparse System Identification with Grokked MLP Libraries

- Paper: https://huggingface.co/pandevim/cs810/tree/main/paper
- Google Colab (experiment code): https://colab.research.google.com/drive/1BI5pjA5e4O-FTY6DUd1JVqxQdXxY5EHP?usp=sharing
- Hugging Face (intermediate weights and checkpoints): https://huggingface.c (o/pandevim/cs810/tree/main
- W&B Dashboard (experiment observations): https://wandb.ai/pandevim-wandb-team/cs810/reports/Neural-SINDy-Experiment-Results--VmlldzoxNjI5NjAwOA


#### References
- The main _SINDy_ paper my idea extend [Discovering governing equations from data: Sparse identification of nonlinear dynamical systems](https://arxiv.org/pdf/1509.03580)
- Library Generation: [GROKKING: GENERALIZATION BEYOND OVERFIT-TING ON SMALL ALGORITHMIC DATASETS](https://arxiv.org/pdf/2201.02177)
- Arithmetics: [Grokking modular arithmetic](https://arxiv.org/pdf/2301.02679)
- Router: [Categorical Reparameterization with Gumbel-Softmax](https://arxiv.org/pdf/1611.01144)

#### Experiment setup
- A100 (80GB VRAM)

In [1]:
%%capture
!pip install pysr

In [18]:
import io
import os

import numpy as np
import pandas as pd

import torch
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F

from scipy.integrate import solve_ivp
import matplotlib.pyplot as plt
import json
from tqdm.auto import tqdm

from sklearn.linear_model import Lasso, Ridge

from google.colab import userdata
from huggingface_hub import HfApi, login, hf_hub_download
import wandb

In [19]:
HF_TOKEN = "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"
WANDB_API_KEY = "REDACTED_WANDB_TOKEN"

In [20]:
os.environ["WANDB_API_KEY"] = WANDB_API_KEY
wandb.login(WANDB_API_KEY)
os.environ["HF_TOKEN"] = HF_TOKEN
login(token=HF_TOKEN)
!hf auth login --token $HF_TOKEN

wandb: WARNING If you're specifying your api key in code, ensure this code is not shared publicly.
wandb: WARNING Consider setting the WANDB_API_KEY environment variable, or running `wandb login` from the command line.
wandb: WARNING [wandb.login()] Changing session credentials to explicit value for https://api.wandb.ai.
wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc
Note: Environment variable`HF_TOKEN` is set and is the current active token independently from the token you've just configured.


The token has not been saved to the git credentials helper. Pass `add_to_git_credential=True` in this function directly or `--add-to-git-credential` if using via `hf`CLI if you want to set the git credential as well.
Token is valid (permission: fineGrained).
The token `HF_TOKEN` has been saved to /root/.cache/huggingface/stored_tokens
Your token has been saved to /root/.cache/huggingface/token
Login successful.
Note: Environment variable`HF_TOKEN` is set and is the current active token independently from the token you've just configured.


In [21]:
api = HfApi()
HF_REPO_ID = "pandevim/cs810"

In [22]:
device = "cuda"
torch.set_default_device(device)

# Phase 1: Train MLPs to "grok" basic mathematical operations

**Grokking** = delayed generalization.  
We train well past overfitting (10k-20k epochs each) using AdamW with **heavy weight decay** (the key ingredient for grokking), and the network eventually restructures its weights into an algorithmic circuit that generalizes perfectly.

### Tasks
We train 5 MLPs on the following functions:
- `cos(x)`
- `sin(x)`
- `identity(x)`
- `x + y`
- `x * y`

## Observation
- Training loss drops fast, validation loss later implying grokking (![Grokking Curves: Train vs Validation Loss](paper/grokking_curves.png)).

## Outputs
- [checkpoints/mlp_*.pt](https://huggingface.co/pandevim/cs810/tree/main/phase1/checkpoints)
- [paper/grokking_curves.png](https://huggingface.co/pandevim/cs810/blob/main/phase1/plots/grokking_curves.png)

In [12]:
# ─── MLP Architecture ────────────────────────────────────────────────────────

class GrokMLP(nn.Module):
    """Small MLP designed to grok a mathematical function."""

    def __init__(self, input_dim=1, hidden_dim=128, num_layers=2, activation="tanh"):
        super().__init__()
        layers = []
        act_fn = nn.Tanh() if activation == "tanh" else nn.ReLU()

        # Input layer
        layers.append(nn.Linear(input_dim, hidden_dim))
        layers.append(act_fn)

        # Hidden layers
        for _ in range(num_layers - 1):
            layers.append(nn.Linear(hidden_dim, hidden_dim))
            layers.append(act_fn)

        # Output layer
        layers.append(nn.Linear(hidden_dim, 1))

        self.net = nn.Sequential(*layers)

    def forward(self, x):
        return self.net(x)

In [ ]:
# ─── Data Generation ─────────────────────────────────────────────────────────

def make_unary_data(func, n_train=500, n_val=200, x_range=(-2 * np.pi, 2 * np.pi)):
    """Generate train/val data for a unary function f(x)."""
    rng = np.random.default_rng(42)

    x_train = rng.uniform(*x_range, size=(n_train, 1)).astype(np.float32)
    y_train = func(x_train).astype(np.float32)

    x_val = rng.uniform(*x_range, size=(n_val, 1)).astype(np.float32)
    y_val = func(x_val).astype(np.float32)

    return (
        torch.from_numpy(x_train), torch.from_numpy(y_train),
        torch.from_numpy(x_val), torch.from_numpy(y_val),
    )


def make_binary_data(func, n_train=1000, n_val=400, x_range=(-5.0, 5.0)):
    """Generate train/val data for a binary function f(x, y)."""
    rng = np.random.default_rng(42)

    xy_train = rng.uniform(*x_range, size=(n_train, 2)).astype(np.float32)
    z_train = func(xy_train[:, 0:1], xy_train[:, 1:2]).astype(np.float32)

    xy_val = rng.uniform(*x_range, size=(n_val, 2)).astype(np.float32)
    z_val = func(xy_val[:, 0:1], xy_val[:, 1:2]).astype(np.float32)

    return (
        torch.from_numpy(xy_train), torch.from_numpy(z_train),
        torch.from_numpy(xy_val), torch.from_numpy(z_val),
    )


# ─── Training Loop ───────────────────────────────────────────────────────────

def train_grok(
    model, x_train, y_train, x_val, y_val,
    epochs=20000, lr=1e-3, weight_decay=1.0,
    log_every=500, device="cuda"
):
    """
    Train with heavy weight decay — the key ingredient for grokking.
    Weight decay forces the network to find a simpler, generalizing solution
    rather than memorizing via large, complex weight patterns.
    """
    model = model.to(device)
    x_train, y_train = x_train.to(device), y_train.to(device)
    x_val, y_val = x_val.to(device), y_val.to(device)

    optimizer = optim.AdamW(model.parameters(), lr=lr, weight_decay=weight_decay)
    scheduler = optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=epochs)
    criterion = nn.MSELoss()

    history = {"train_loss": [], "val_loss": [], "epoch": []}
    run = wandb.run

    with tqdm(range(1, epochs + 1), desc="Training", leave=True) as pbar:
      for epoch in pbar:
          # ── Train ──
          model.train()
          pred = model(x_train)
          loss = criterion(pred, y_train)
          optimizer.zero_grad()
          loss.backward()
          optimizer.step()
          scheduler.step()

          # ── Validate ──
          if epoch % log_every == 0 or epoch == 1:
              model.eval()
              with torch.no_grad():
                  val_pred = model(x_val)
                  val_loss = criterion(val_pred, y_val)

              pbar.set_postfix({"train": f"{loss.item():.4f}", "val": f"{val_loss.item():.4f}"})

              history["train_loss"].append(loss.item())
              history["val_loss"].append(val_loss.item())
              history["epoch"].append(epoch)

              wandb.log({"train_loss": loss.item(), "val_loss": val_loss.item(), "epoch": epoch})
              if run and run.disabled:
                  print("\n⚠️  Run stopped from W&B dashboard.")
                  break

      return history


# ─── Main ─────────────────────────────────────────────────────────────────────
# ── Define the functions to grok ──
tasks = {
    "cos": {
        "func": lambda x: np.cos(x),
        "input_dim": 1,
        "activation": "tanh",
        "data_fn": lambda f: make_unary_data(f),
        "epochs": 20000,
        "weight_decay": 1.0,
        "lr": 1e-3,
    },
    "sin": {
        "func": lambda x: np.sin(x),
        "input_dim": 1,
        "activation": "tanh",
        "data_fn": lambda f: make_unary_data(f),
        "epochs": 20000,
        "weight_decay": 1.0,
        "lr": 1e-3,
    },
    "add": {
        "func": lambda x, y: x + y,
        "input_dim": 2,
        "activation": "relu",
        "data_fn": lambda f: make_binary_data(f),
        "epochs": 15000,
        "weight_decay": 0.5,
        "lr": 1e-3,
    },
    "mul": {
        "func": lambda x, y: x * y,
        "input_dim": 2,
        "activation": "relu",
        "data_fn": lambda f: make_binary_data(f),
        "epochs": 20000,
        "weight_decay": 0.5,
        "lr": 1e-3,
    },
    "identity": {
        "func": lambda x: x,
        "input_dim": 1,
        "activation": "relu",
        "data_fn": lambda f: make_unary_data(f, x_range=(-5.0, 5.0)),
        "epochs": 10000,
        "weight_decay": 1.0,
        "lr": 1e-3,
    },
}

all_histories = {}

for name, cfg in tasks.items():
    print(f"{'='*60}")
    print(f"  Training MLP_{name}")
    print(f"{'='*60}")

    # Build model
    model = GrokMLP(
        input_dim=cfg["input_dim"],
        hidden_dim=128,
        num_layers=2,
        activation=cfg["activation"],
    )

    # Generate data
    x_train, y_train, x_val, y_val = cfg["data_fn"](cfg["func"])
    wandb.init(
        project="cs810",
        name=f"MLP_{name}",
        config={
            "epochs": cfg["epochs"],
            "lr": cfg["lr"],
            "weight_decay": cfg["weight_decay"],
            "activation": cfg["activation"],
            "input_dim": cfg["input_dim"],
            "hidden_dim": 128,
            "num_layers": 2,
        },
        reinit="finish_previous",
    )
    wandb.watch(model, log="all", log_freq=500)

    # Train
    history = train_grok(
        model, x_train, y_train, x_val, y_val,
        epochs=cfg["epochs"],
        lr=cfg["lr"],
        weight_decay=cfg["weight_decay"],
        device=device,
    )

    all_histories[name] = history
    wandb.finish()

    # Save model
    buf = io.BytesIO()
    torch.save({"model_state_dict": model.state_dict(), "input_dim": cfg["input_dim"],
                "activation": cfg["activation"], "hidden_dim": 128, "num_layers": 2}, buf)
    api.upload_file(path_or_fileobj=io.BytesIO(buf.getvalue()),
                    path_in_repo=f"phase1/checkpoints/mlp_{name}.pt",
                    repo_id=HF_REPO_ID, commit_message=f"Add mlp_{name}")

    # Final validation error
    model.eval()
    with torch.no_grad():
        x_val_dev = x_val.to(device)
        y_val_dev = y_val.to(device)
        val_pred = model(x_val_dev)
        final_mse = nn.MSELoss()(val_pred, y_val_dev).item()
        # Relative error
        y_range = (y_val_dev.max() - y_val_dev.min()).item()
        rel_err = np.sqrt(final_mse) / max(y_range, 1e-8) * 100
        print(f"\n  ✓ MLP_{name} — Final Val MSE: {final_mse:.6f}, "
              f"Relative Error: {rel_err:.2f}%\n")

# ── Plot grokking curves ──
fig, axes = plt.subplots(2, 3, figsize=(15, 8))
axes = axes.flatten()
for idx, (name, hist) in enumerate(all_histories.items()):
    ax = axes[idx]
    ax.semilogy(hist["epoch"], hist["train_loss"], label="Train", alpha=0.8)
    ax.semilogy(hist["epoch"], hist["val_loss"], label="Val", alpha=0.8)
    ax.set_title(f"MLP_{name}", fontsize=12, fontweight="bold")
    ax.set_xlabel("Epoch")
    ax.set_ylabel("MSE Loss (log)")
    ax.legend()
    ax.grid(True, alpha=0.3)

# Hide unused subplot
if len(all_histories) < len(axes):
    for j in range(len(all_histories), len(axes)):
        axes[j].set_visible(False)

plt.suptitle("Grokking Curves: Train vs Validation Loss", fontsize=14, fontweight="bold")
plt.tight_layout()
buf = io.BytesIO()
plt.savefig(buf, format="png", dpi=150, bbox_inches="tight")
api.upload_file(path_or_fileobj=io.BytesIO(buf.getvalue()),
                path_in_repo="paper/grokking_curves.png",
                repo_id=HF_REPO_ID, commit_message="Add grokking curves plot")
plt.close()
print(f"\n✓ Grokking curves saved to paper/grokking_curves.png")

# Save histories
api.upload_file(path_or_fileobj=io.BytesIO(json.dumps(all_histories, indent=2).encode()),
                path_in_repo="phase1/checkpoints/training_histories.json",
                repo_id=HF_REPO_ID, commit_message="Add training histories")

print("\n✓ All models saved to checkpoints/")
print("✓ Done!")

  Training MLP_cos


Training:   0%|          | 0/20000 [00:00<?, ?it/s]

KeyboardInterrupt: 

# Phase 2: Generate Synthetic ODE Data for Testing

We use a **damped harmonic oscillator**:

$$
\ddot{x} = -k \cdot x - c \cdot \dot{x}
$$

i.e: `ẍ = -1.0·x - 0.1·ẋ` (600 time points, small Gaussian noise added)

### First-Order System Form

Rewriting as a system of first-order differential equations:

$$
\frac{dx}{dt} = v
$$

$$
\frac{dv}{dt} = -k \cdot x - c \cdot v
$$

### Key Insight

This system involves **linear combinations of state variables**, which is exactly what our MLP library should be able to discover.

## Outputs
- Time series `[t, x, v, dx/dt, dv/dt]`: [data/oscillator_data.npz](https://huggingface.co/pandevim/cs810/blob/main/phase1/data/oscillator_data.npz)
- Phase portrait and time series: [paper/oscillator_data.png](https://huggingface.co/pandevim/cs810/blob/main/phase1/plots/oscillator_data.png)

In [13]:
def generate_damped_oscillator(
    k=1.0, c=0.1,
    x0=1.0, v0=0.0,
    t_span=(0, 30), dt=0.05,
    noise_std=0.001,
    seed=42,
):
    """
    Generate time-series data from a damped harmonic oscillator.

    Returns:
        t: time points (N,)
        X: state matrix [x, v] of shape (N, 2)
        dXdt: derivatives [dx/dt, dv/dt] of shape (N, 2)
        params: dict with k, c
    """
    rng = np.random.default_rng(seed)

    def rhs(t, state):
        x, v = state
        dxdt = v
        dvdt = -k * x - c * v
        return [dxdt, dvdt]

    t_eval = np.arange(t_span[0], t_span[1], dt)
    sol = solve_ivp(rhs, t_span, [x0, v0], t_eval=t_eval, method="RK45", max_step=dt/2)

    t = sol.t
    X = sol.y.T  # (N, 2) — columns are [x, v]

    # Compute clean derivatives
    dXdt = np.zeros_like(X)
    dXdt[:, 0] = X[:, 1]                    # dx/dt = v
    dXdt[:, 1] = -k * X[:, 0] - c * X[:, 1]  # dv/dt = -k·x - c·v

    # Add measurement noise to state (NOT to derivatives)
    X_noisy = X + rng.normal(0, noise_std, size=X.shape)

    return t, X_noisy, dXdt, {"k": k, "c": c}

In [ ]:
print("Generating damped harmonic oscillator data...")
print("  ẍ = -k·x - c·ẋ  with k=1.0, c=0.1")
print("  x(0) = 1.0, ẋ(0) = 0.0\n")

t, X, dXdt, params = generate_damped_oscillator()

print(f"  Time points:  {len(t)}")
print(f"  State shape:  {X.shape}")
print(f"  dX/dt shape:  {dXdt.shape}")
print(f"  Parameters:   k={params['k']}, c={params['c']}")

# Save data
buf = io.BytesIO()
np.savez(buf, t=t, X=X, dXdt=dXdt, k=params["k"], c=params["c"])
api.upload_file(path_or_fileobj=io.BytesIO(buf.getvalue()),
                path_in_repo="phase1/data/oscillator_data.npz",
                repo_id=HF_REPO_ID, commit_message="Add oscillator data")
print("\n✓ Data uploaded to HuggingFace")

# Plot
fig, axes = plt.subplots(2, 2, figsize=(12, 8))

# Phase portrait
axes[0, 0].plot(X[:, 0], X[:, 1], "b-", alpha=0.7, linewidth=0.8)
axes[0, 0].plot(X[0, 0], X[0, 1], "go", markersize=8, label="Start")
axes[0, 0].plot(X[-1, 0], X[-1, 1], "rs", markersize=8, label="End")
axes[0, 0].set_xlabel("x (position)")
axes[0, 0].set_ylabel("v (velocity)")
axes[0, 0].set_title("Phase Portrait")
axes[0, 0].legend()
axes[0, 0].grid(True, alpha=0.3)

# Time series - position
axes[0, 1].plot(t, X[:, 0], "b-", alpha=0.8, label="x(t)")
axes[0, 1].set_xlabel("Time")
axes[0, 1].set_ylabel("Position x")
axes[0, 1].set_title("Position vs Time")
axes[0, 1].grid(True, alpha=0.3)

# Time series - velocity
axes[1, 0].plot(t, X[:, 1], "r-", alpha=0.8, label="v(t)")
axes[1, 0].set_xlabel("Time")
axes[1, 0].set_ylabel("Velocity v")
axes[1, 0].set_title("Velocity vs Time")
axes[1, 0].grid(True, alpha=0.3)

# Derivatives
axes[1, 1].plot(t, dXdt[:, 0], "b-", alpha=0.6, label="dx/dt")
axes[1, 1].plot(t, dXdt[:, 1], "r-", alpha=0.6, label="dv/dt")
axes[1, 1].set_xlabel("Time")
axes[1, 1].set_ylabel("Derivative")
axes[1, 1].set_title("True Derivatives")
axes[1, 1].legend()
axes[1, 1].grid(True, alpha=0.3)

plt.suptitle(
    f"Damped Harmonic Oscillator: ẍ = -{params['k']}·x - {params['c']}·ẋ",
    fontsize=14, fontweight="bold"
)
plt.tight_layout()
buf = io.BytesIO()
plt.savefig(buf, format="png", dpi=150, bbox_inches="tight")
plt.close()
api.upload_file(path_or_fileobj=io.BytesIO(buf.getvalue()),
                path_in_repo="phase1/plots/oscillator_data.png",
                repo_id=HF_REPO_ID, commit_message="Add oscillator plot")
print("✓ Plot uploaded to HuggingFace")
print("✓ Done!")

Generating damped harmonic oscillator data...
  ẍ = -k·x - c·ẋ  with k=1.0, c=0.1
  x(0) = 1.0, ẋ(0) = 0.0

  Time points:  600
  State shape:  (600, 2)
  dX/dt shape:  (600, 2)
  Parameters:   k=1.0, c=0.1


Uploading files as a binary IO buffer is not supported by Xet Storage. Falling back to HTTP upload.



✓ Data uploaded to HuggingFace


Uploading files as a binary IO buffer is not supported by Xet Storage. Falling back to HTTP upload.


✓ Plot uploaded to HuggingFace
✓ Done!


# Phase 3: Neural SINDy — Using Grokked MLPs as Basis Functions

**Standard SINDy**:

$$
\dot{X} = \Theta(X) \cdot \xi
$$

- Where:  
  $$
  \Theta = [1, x, x^2, \sin(x), \cos(x), \dots]
  $$  
  (hand-crafted library)  
  $\xi$ is a sparse coefficient vector

**Neural SINDy**:

$$
\dot{X} = \Theta_{\text{neural}}(X) \cdot \xi
$$

- Where:  
  $$ \Theta_{\text{neural}} = [\{\text{MLP}_{\text{identity}}\}(x), \{\text{MLP}_{\text{cos}}\}(x), \{\text{MLP}_{\text{sin}}\}(x), \dots] $$
  (grokked MLP library)  
  $\xi$ is a sparse coefficient vector found via STLSQ or LASSO

**Key Insight:**  
Because each MLP is a differentiable function learned to approximate a clean mathematical operation, the sparse regression selects which *operations* best explain the dynamics.

In [14]:
# ─── MLP Library ─────────────────────────────────────────────────────────────

class MLPLibraryTerm:
    """A single term in the neural SINDy library."""

    def __init__(self, name, model, input_dim, input_columns):
        """
        Args:
            name: human-readable name (e.g., "cos(x)")
            model: trained GrokMLP
            input_dim: how many inputs the MLP expects (1 or 2)
            input_columns: which columns of the state X to feed in.
                           For unary: [0] or [1]
                           For binary: [0, 1]
        """
        self.name = name
        self.model = model
        self.input_dim = input_dim
        self.input_columns = input_columns

    def evaluate(self, X):
        """
        Evaluate this MLP on state data X of shape (N, state_dim).
        Returns (N, 1) array.
        """
        self.model.eval()
        with torch.no_grad():
            # Select relevant columns
            x_in = torch.from_numpy(
                X[:, self.input_columns].astype(np.float32)
            )
            out = self.model(x_in).numpy()
        return out


class NeuralSINDyLibrary:
    """
    Library of grokked MLPs for SINDy-style system identification.

    Given state data X (shape N x state_dim), evaluates every MLP
    in the library to construct the feature matrix Θ.
    """

    def __init__(self):
        self.terms = []

    def add_term(self, name, model, input_dim, input_columns):
        term = MLPLibraryTerm(name, model, input_dim, input_columns)
        self.terms.append(term)
        return self

    def build_theta(self, X):
        """
        Construct the library matrix Θ of shape (N, n_terms).

        Each column is the output of one MLP evaluated on the state data.
        """
        columns = []
        for term in self.terms:
            col = term.evaluate(X)  # (N, 1)
            columns.append(col)

        Theta = np.hstack(columns)  # (N, n_terms)
        return Theta

    def get_term_names(self):
        return [t.name for t in self.terms]


# ─── Sparse Regression (STLSQ) ──────────────────────────────────────────────

def stlsq(Theta, dXdt_col, threshold=0.05, max_iter=20, alpha=0.01):
    """
    Sequentially Thresholded Least Squares (STLSQ).

    This is the original SINDy optimizer:
    1. Solve least squares: ξ = (Θ^T Θ)^-1 Θ^T Ẋ
    2. Zero out coefficients below threshold
    3. Repeat with remaining terms until stable

    Args:
        Theta: library matrix (N, n_terms)
        dXdt_col: single column of derivatives (N,)
        threshold: sparsity cutoff
        max_iter: max iterations
        alpha: ridge regularization

    Returns:
        xi: sparse coefficient vector (n_terms,)
    """
    n_terms = Theta.shape[1]
    xi = np.zeros(n_terms)

    # Initial ridge regression
    ridge = Ridge(alpha=alpha, fit_intercept=False)
    ridge.fit(Theta, dXdt_col)
    xi = ridge.coef_.copy()

    for iteration in range(max_iter):
        # Threshold small coefficients
        small_idx = np.abs(xi) < threshold
        xi[small_idx] = 0.0

        # Re-fit with remaining terms
        big_idx = ~small_idx
        if not np.any(big_idx):
            break

        ridge_sub = Ridge(alpha=alpha, fit_intercept=False)
        ridge_sub.fit(Theta[:, big_idx], dXdt_col)
        xi[big_idx] = ridge_sub.coef_

        # Check convergence
        if np.all(np.abs(xi[small_idx]) == 0):
            # Check if any newly fitted coefficient is below threshold
            if np.all(np.abs(xi[big_idx]) >= threshold):
                break

    return xi


# ─── Load Grokked Models ────────────────────────────────────────────────────

def load_grokked_mlp(checkpoint_path):
    """Load a trained GrokMLP from checkpoint."""
    ckpt = torch.load(checkpoint_path, map_location="cpu", weights_only=True)
    model = GrokMLP(
        input_dim=ckpt["input_dim"],
        hidden_dim=ckpt["hidden_dim"],
        num_layers=ckpt["num_layers"],
        activation=ckpt["activation"],
    )
    model.load_state_dict(ckpt["model_state_dict"])
    model.eval()
    return model, ckpt["input_dim"]


def build_default_library(repo_id="pandevim/cs810"):
    """
    Build a NeuralSINDyLibrary from all trained MLP checkpoints.

    For a 2D state [x, v], we create terms:
        - Unary MLPs applied to x (column 0)
        - Unary MLPs applied to v (column 1)
        - Binary MLPs applied to (x, v)
        - Constant bias term
    """
    library = NeuralSINDyLibrary()

    # Load all models
    unary_names = ["identity", "cos", "sin"]
    binary_names = ["add", "mul"]

    for name in unary_names:
        try:
            path = hf_hub_download(repo_id=repo_id, filename=f"phase1/checkpoints/mlp_{name}.pt")
        except Exception:
            print(f"  ⚠ Checkpoint not found on HF: mlp_{name}.pt, skipping")
            continue

        model, input_dim = load_grokked_mlp(path)

        # Apply to each state variable
        library.add_term(f"{name}(x)", model, input_dim, [0])
        library.add_term(f"{name}(v)", model, input_dim, [1])

    for name in binary_names:
        try:
            path = hf_hub_download(repo_id=repo_id, filename=f"phase1/checkpoints/mlp_{name}.pt")
        except Exception:
            print(f"  ⚠ Checkpoint not found on HF: mlp_{name}.pt, skipping")
            continue

        model, input_dim = load_grokked_mlp(path)
        library.add_term(f"{name}(x,v)", model, input_dim, [0, 1])

    return library


# ─── Discovery ───────────────────────────────────────────────────────────────

def discover_equations(X, dXdt, library, threshold=0.05, alpha=0.01):
    """
    Run Neural SINDy discovery.

    Args:
        X: state data (N, state_dim)
        dXdt: derivatives (N, state_dim)
        library: NeuralSINDyLibrary instance
        threshold: STLSQ sparsity threshold
        alpha: ridge regularization

    Returns:
        xi: coefficient matrix (n_terms, state_dim)
        term_names: list of term names
    """
    print("\n  Building library matrix Θ...")
    Theta = library.build_theta(X)
    term_names = library.get_term_names()
    n_terms = len(term_names)
    state_dim = dXdt.shape[1]

    print(f"  Θ shape: {Theta.shape} ({len(term_names)} library terms)")
    print(f"  Library terms: {term_names}\n")

    xi = np.zeros((n_terms, state_dim))

    state_labels = ["ẋ (dx/dt)", "v̇ (dv/dt)"]

    for col in range(state_dim):
        print(f"  ─── Discovering equation for {state_labels[col]} ───")
        xi[:, col] = stlsq(Theta, dXdt[:, col], threshold=threshold, alpha=alpha)

        # Print discovered equation
        active = np.abs(xi[:, col]) > 1e-10
        terms = []
        for i in range(n_terms):
            if active[i]:
                coeff = xi[i, col]
                terms.append(f"{coeff:+.4f}·{term_names[i]}")

        eq_str = " ".join(terms) if terms else "0"
        print(f"  {state_labels[col]} = {eq_str}\n")

    return xi, term_names

# Phase 4: Run the Full Neural SINDy Discovery Experiment

### Pipeline

1. **Load Grokked MLPs from Checkpoints**
2. **Load ODE Data**
3. **Build Library Matrix Θ** by evaluating each MLP on state data
4. **Solve Sparse Regression**
    - Fit the dynamics using sparse regression: $$ \dot{X} \approx \Theta(X) \cdot \xi $$
    - Determine which MLPs contribute to the dynamics and their coefficients **ξ**.
      * Tries multiple sparsity thresholds, picks the sparsest model with low error
5. **Report Selected MLPs and Coefficients**
6. **Validate Discovered System**
   - Simulate the system using the learned ξ coefficients.
   - Compare the simulation against the ground-truth ODE trajectories to assess accuracy.

### Outputs
- [data/discovery_results.npz](https://huggingface.co/pandevim/cs810/blob/main/phase1/data/discovery_results.npz)
- [paper/discovery_comparison.png](https://huggingface.co/pandevim/cs810/blob/main/phase1/plots/discovery_comparison.png)
- [paper/coefficients.png](https://huggingface.co/pandevim/cs810/blob/main/phase1/plots/coefficients.png)

In [15]:
def simulate_discovered_system(xi, term_names, library, x0, t_span, dt):
    """
    Simulate the discovered system forward in time.

    The discovered dynamics are: Ẋ = Θ(X) · ξ
    We numerically integrate this ODE to produce trajectories.
    """
    def rhs(t, state):
        X = state.reshape(1, -1)  # (1, state_dim)
        Theta = library.build_theta(X)  # (1, n_terms)
        dXdt = (Theta @ xi).flatten()  # (state_dim,)
        return dXdt

    t_eval = np.arange(t_span[0], t_span[1], dt)
    sol = solve_ivp(rhs, t_span, x0, t_eval=t_eval, method="RK45", max_step=dt)
    return sol.t, sol.y.T

In [ ]:
# ── Load data ──
print("Loading oscillator data...")
path = hf_hub_download(repo_id=HF_REPO_ID, filename="phase1/data/oscillator_data.npz")
data = np.load(path)
t = data["t"]
X = data["X"]
dXdt = data["dXdt"]
k_true = float(data["k"])
c_true = float(data["c"])
print(f"  Loaded {len(t)} time points, true params: k={k_true}, c={c_true}\n")

# ── Build library ──
print("Building Neural SINDy library from grokked MLPs...")
library = build_default_library(HF_REPO_ID)

# ── Discover equations ──
print("\n" + "=" * 60)
print("  NEURAL SINDy DISCOVERY")
print("=" * 60)

# Try a range of thresholds and pick the sparsest good one
best_xi = None
best_threshold = None
best_error = float("inf")

for threshold in [0.01, 0.025, 0.05, 0.1, 0.2]:
    xi, term_names = discover_equations(
        X, dXdt, library,
        threshold=threshold,
        alpha=0.01,
    )

    # Evaluate reconstruction error
    Theta = library.build_theta(X)
    dXdt_pred = Theta @ xi
    mse = np.mean((dXdt - dXdt_pred) ** 2)
    n_active = np.sum(np.abs(xi) > 1e-10)

    print(f"  Threshold={threshold:.3f}: MSE={mse:.6f}, "
          f"Active terms={n_active}/{xi.size}")

    # Pick the sparsest model with acceptable error
    if mse < 0.01 and (best_xi is None or n_active <= np.sum(np.abs(best_xi) > 1e-10)):
        best_xi = xi.copy()
        best_threshold = threshold
        best_error = mse

if best_xi is None:
    # Fall back to lowest error
    best_xi = xi
    best_threshold = threshold
    best_error = mse
    print("\n  ⚠ Could not find a sparse model with low error, using last result")

xi = best_xi
print(f"\n  ✓ Best threshold: {best_threshold}, MSE: {best_error:.6f}")

# ── Print final discovered equations ──
print("\n" + "=" * 60)
print("  DISCOVERED EQUATIONS")
print("=" * 60)

state_labels = ["ẋ = dx/dt", "v̇ = dv/dt"]
for col in range(dXdt.shape[1]):
    active = np.abs(xi[:, col]) > 1e-10
    terms = []
    for i in range(len(term_names)):
        if active[i]:
            coeff = xi[i, col]
            terms.append(f"{coeff:+.4f} · {term_names[i]}")

    eq_str = " ".join(terms) if terms else "0"
    print(f"\n  {state_labels[col]} = {eq_str}")

print(f"\n  TRUE EQUATIONS:")
print(f"  ẋ = +1.0 · v")
print(f"  v̇ = -{k_true:.1f} · x  -{c_true:.1f} · v")

# ── Save results ──
buf = io.BytesIO()
np.savez(buf, xi=xi, term_names=np.array(term_names), best_threshold=best_threshold, mse=best_error)
api.upload_file(path_or_fileobj=io.BytesIO(buf.getvalue()),
                path_in_repo="phase1/data/discovery_results.npz",
                repo_id=HF_REPO_ID, commit_message="Add discovery results")

# ── Simulate discovered system ──
print("\n\nSimulating discovered system...")
try:
    t_sim, X_sim = simulate_discovered_system(
        xi, term_names, library,
        x0=X[0],
        t_span=(t[0], t[-1]),
        dt=0.05,
    )

    # Plot comparison
    fig, axes = plt.subplots(1, 2, figsize=(12, 5))

    axes[0].plot(t, X[:, 0], "b-", alpha=0.8, label="True x(t)")
    axes[0].plot(t_sim, X_sim[:, 0], "r--", alpha=0.8, label="Discovered x(t)")
    axes[0].set_xlabel("Time")
    axes[0].set_ylabel("Position x")
    axes[0].set_title("Position: True vs Discovered")
    axes[0].legend()
    axes[0].grid(True, alpha=0.3)

    axes[1].plot(t, X[:, 1], "b-", alpha=0.8, label="True v(t)")
    axes[1].plot(t_sim, X_sim[:, 1], "r--", alpha=0.8, label="Discovered v(t)")
    axes[1].set_xlabel("Time")
    axes[1].set_ylabel("Velocity v")
    axes[1].set_title("Velocity: True vs Discovered")
    axes[1].legend()
    axes[1].grid(True, alpha=0.3)

    plt.suptitle("Neural SINDy: True vs Discovered Dynamics", fontsize=14, fontweight="bold")
    plt.tight_layout()
    buf = io.BytesIO()
    plt.savefig(buf, format="png", dpi=150, bbox_inches="tight")
    plt.close()
    api.upload_file(path_or_fileobj=io.BytesIO(buf.getvalue()),
                    path_in_repo="phase1/plots/discovery_comparison.png",
                    repo_id=HF_REPO_ID, commit_message="Add discovery comparison plot")
    print("  ✓ Comparison plot uploaded to HuggingFace")
except Exception as e:
    print(f"  ⚠ Simulation failed (this is OK for PoC): {e}")

# ── Coefficient bar chart ──
fig, axes = plt.subplots(1, 2, figsize=(14, 5))
x_pos = np.arange(len(term_names))

for col, (ax, label) in enumerate(zip(axes, state_labels)):
    colors = ["#2196F3" if abs(v) > 1e-10 else "#BDBDBD" for v in xi[:, col]]
    ax.bar(x_pos, xi[:, col], color=colors, edgecolor="black", linewidth=0.5)
    ax.set_xticks(x_pos)
    ax.set_xticklabels(term_names, rotation=45, ha="right", fontsize=9)
    ax.set_ylabel("Coefficient ξ")
    ax.set_title(f"Sparse Coefficients for {label}")
    ax.axhline(y=0, color="black", linewidth=0.5)
    ax.grid(True, alpha=0.3, axis="y")

plt.suptitle("Neural SINDy: Selected Library Terms", fontsize=14, fontweight="bold")
plt.tight_layout()
buf = io.BytesIO()
plt.savefig(buf, format="png", dpi=150, bbox_inches="tight")
plt.close()
api.upload_file(path_or_fileobj=io.BytesIO(buf.getvalue()),
                path_in_repo="phase1/plots/coefficients.png",
                repo_id=HF_REPO_ID, commit_message="Add coefficients plot")
print("  ✓ Coefficient bar chart uploaded to HuggingFace")

print("\n✓ Experiment complete!")

Loading oscillator data...


phase1/data/oscillator_data.npz:   0%|          | 0.00/25.2k [00:00<?, ?B/s]

  Loaded 600 time points, true params: k=1.0, c=0.1

Building Neural SINDy library from grokked MLPs...


phase1/checkpoints/mlp_identity.pt:   0%|          | 0.00/70.5k [00:00<?, ?B/s]

phase1/checkpoints/mlp_cos.pt:   0%|          | 0.00/70.5k [00:00<?, ?B/s]

phase1/checkpoints/mlp_sin.pt:   0%|          | 0.00/70.5k [00:00<?, ?B/s]

phase1/checkpoints/mlp_add.pt:   0%|          | 0.00/71.0k [00:00<?, ?B/s]

phase1/checkpoints/mlp_mul.pt:   0%|          | 0.00/71.0k [00:00<?, ?B/s]


  NEURAL SINDy DISCOVERY

  Building library matrix Θ...
  Θ shape: (600, 8) (8 library terms)
  Library terms: ['identity(x)', 'identity(v)', 'cos(x)', 'cos(v)', 'sin(x)', 'sin(v)', 'add(x,v)', 'mul(x,v)']

  ─── Discovering equation for ẋ (dx/dt) ───
  ẋ (dx/dt) = -0.3172·identity(x) +0.6502·identity(v) -0.0123·sin(x) +0.0212·sin(v) +0.3291·add(x,v)

  ─── Discovering equation for v̇ (dv/dt) ───
  v̇ (dv/dt) = -0.6208·identity(x) +0.2621·identity(v) -0.0175·sin(x) -0.3621·add(x,v)

  Threshold=0.010: MSE=0.000001, Active terms=9/16

  Building library matrix Θ...
  Θ shape: (600, 8) (8 library terms)
  Library terms: ['identity(x)', 'identity(v)', 'cos(x)', 'cos(v)', 'sin(x)', 'sin(v)', 'add(x,v)', 'mul(x,v)']

  ─── Discovering equation for ẋ (dx/dt) ───
  ẋ (dx/dt) = -0.3325·identity(x) +0.6675·identity(v) +0.3324·add(x,v)

  ─── Discovering equation for v̇ (dv/dt) ───
  v̇ (dv/dt) = -0.6322·identity(x) +0.2677·identity(v) -0.3676·add(x,v)

  Threshold=0.025: MSE=0.000001, Active 

Uploading files as a binary IO buffer is not supported by Xet Storage. Falling back to HTTP upload.




Simulating discovered system...


Uploading files as a binary IO buffer is not supported by Xet Storage. Falling back to HTTP upload.


  ✓ Comparison plot uploaded to HuggingFace
  ✓ Coefficient bar chart uploaded to HuggingFace

✓ Experiment complete!


# Analysis from Phase 4
**Result: ✅ Mathematically correct, but coefficients are spread across redundant terms.**

Since `add(x,v) ≈ x + v`, substituting and simplifying:

| Equation | Discovered (simplified) | True | Match? |
|----------|------------------------|------|--------|
| ẋ | (-0.3325 + 0.3324)·x + (0.6675 + 0.3324)·v = **-0.0001·x + 0.9999·v** | +1.0·v | ✅ |
| v̇ | (-0.6322 - 0.3676)·x + (0.2677 - 0.3676)·v = **-0.9998·x - 0.0999·v** | -1.0·x - 0.1·v | ✅ |

**MSE:** 0.000001 (essentially exact fit)  
**Active terms:** 6/16

### Key Finding: Collinearity Problem

The $\text{MLP}_{\text{add}}(x, v) = x + v$ is mathematically identical to $\text{MLP}_{\text{identity}}(x) + \text{MLP}_{\text{identity}}(v)$, creating a rank-deficient library matrix as it makes the column $\text{MLP}_{\text{add}}(x, v)$ in $\Theta(X)$ lie in the span of two other columns $\Theta^T \Theta$. STLSQ (even with ridge α=0.01) cannot distinguish between them, so it distributes the coefficients arbitrarily across the three redundant columns. The *combined* answer is correct, but the individual coefficients are uninterpretable. $\text{MLP}_{\text{mul}}(x, v) = x \cdot v$ is genuinely nonlinear, so it is not redundant and should stay.

### Implication
The library design must avoid including terms that are linear combinations of other terms. For the next run, removing $\text{MLP}_{\text{add}}$ from the library (or adding a collinearity detection step) should produce clean, interpretable coefficients.

### What to change
1. **Primary fix (In Phase 3)**: drop $\text{MLP}_{\text{add}}$ from `binary_names` in `build_default_library`. This removes the redundancy at the source and the discovered coefficients will land on $\text{MLP}_{\text{identity}}(x)$ and $\text{MLP}_{\text{identity}}(v)$ directly, clean and interpretable.

2. **Secondary safeguard (optional):**  
Before STLSQ, run a collinearity check on $\Theta(X)$ so the issue surfaces automatically if new redundant basis functions are added later. Two cheap options:
    - **Rank / condition-number check:** If $\mathrm{cond}(\Theta) > 10^8$ or $\mathrm{rank}(\Theta) < n_{\text{terms}}$, raise a warning and drop the offending column (e.g., the column with the highest pairwise $|\mathrm{corr}|$ with another column, or aligned with the smallest singular direction).
    - **Pairwise correlation prune:** Drop any column with $|\mathrm{corr}| > 0.999$ relative to a lower-indexed column.

### Revelation
**Gumbel-Softmax** sidesteps the root cause. STLSQ fails on rank-deficient $\Theta(X)$ because it solves a linear system. A *Gumbel-Softmax router* picks terms via a differentiable argmax over a learned categorical, there is no _normal-equations inversion_, so rank deficiency doesn't produce the "spread across redundant columns" pathology.


# Phase 5: The Gumbel-Softmax Router experiment

This is the end-to-end differentiable variant of Neural SINDy.
Compare results with Phase 4 (STLSQ baseline).

### Pipeline:
1. Load grokked MLPs (frozen)
2. Load ODE data
3. Train a router network with Gumbel-Softmax to learn which MLP
   explains each component of the dynamics
4. Anneal temperature: (τ=5.0, soft selection, exploratory) → (τ=0.05, hard selection, discrete)
  - Uses Straight-Through Estimator: forward pass routes through 1 MLP, backward pass gets smooth gradients
5. Report which MLPs were selected and their learned coefficients
6. Compare against STLSQ baseline results

### Outputs
- [paper/router_results.png](https://huggingface.co/pandevim/cs810/blob/main/phase1/plots/router_results.png)

**What to look for:**

- Training loss should decrease smoothly as temperature anneals
- Final routing should converge to one-hot (one MLP per derivative)
- Should select the same MLPs as STLSQ with similar coefficients

In [16]:
"""
Gumbel-Softmax Neural Router for MLP Library Selection.

Instead of using STLSQ sparse regression (a post-hoc, non-gradient method),
we train a "router" network that learns to select which grokked MLP
to apply at each state, using Gumbel-Softmax for differentiable discrete choices.

Architecture:
    State [x, v] → Router Network → Gumbel-Softmax → Gate (one-hot) → Σ(gate_i · MLP_i(state))

The router learns WHICH MLP explains each component of the dynamics,
while the grokked MLPs provide the actual function approximations.

Key technique: Straight-Through Estimator (STE)
    Forward pass: hard one-hot selection (only 1 MLP computes)
    Backward pass: smooth Gumbel-Softmax gradients flow through all MLPs
"""

# ─── Gumbel-Softmax ──────────────────────────────────────────────────────────

def gumbel_softmax(logits, temperature=1.0, hard=True):
    """
    Sample from the Gumbel-Softmax distribution.

    Args:
        logits: (batch, n_choices) unnormalized log-probabilities
        temperature: τ — controls sharpness. High=soft, Low=hard.
        hard: if True, use Straight-Through Estimator (STE)
              Forward: hard one-hot. Backward: soft gradients.

    Returns:
        (batch, n_choices) — one-hot if hard=True, soft probabilities otherwise
    """
    # Sample Gumbel noise: g_i = -log(-log(u_i)), u_i ~ Uniform(0,1)
    gumbel_noise = -torch.log(-torch.log(
        torch.rand_like(logits).clamp(min=1e-10)
    ))

    # Add noise to logits and apply temperature-scaled softmax
    y_soft = F.softmax((logits + gumbel_noise) / temperature, dim=-1)

    if hard:
        # Straight-Through Estimator
        # Forward: hard one-hot (argmax)
        index = y_soft.max(dim=-1, keepdim=True)[1]
        y_hard = torch.zeros_like(logits).scatter_(-1, index, 1.0)
        # Backward: pretend we used y_soft (gradient flows through)
        return y_hard - y_soft.detach() + y_soft
    else:
        return y_soft


# ─── Router Network ──────────────────────────────────────────────────────────

class GumbelRouter(nn.Module):
    """
    A router that learns to select which grokked MLP to apply for each
    component of the dynamics.

    For a system with state_dim=2 (x, v), we learn TWO routing decisions:
        - Which MLP combination explains dx/dt?
        - Which MLP combination explains dv/dt?

    Each routing decision selects from the full MLP library.
    """

    def __init__(self, state_dim, mlp_modules, mlp_configs, router_hidden=64):
        """
        Args:
            state_dim: dimension of the state space (e.g., 2 for [x, v])
            mlp_modules: list of (name, model, input_columns) tuples
            mlp_configs: list of dicts with 'input_dim' for each MLP
            router_hidden: hidden dimension for the router network
        """
        super().__init__()
        self.state_dim = state_dim
        self.n_mlps = len(mlp_modules)
        self.mlp_names = [m[0] for m in mlp_modules]
        self.mlp_models = nn.ModuleList([m[1] for m in mlp_modules])
        self.input_columns = [m[2] for m in mlp_modules]

        # Freeze grokked MLPs — we don't want to un-grok them!
        for model in self.mlp_models:
            for param in model.parameters():
                param.requires_grad = False

        # One router per state derivative (dx/dt, dv/dt, ...)
        # Each router outputs logits over the MLP library
        self.routers = nn.ModuleList([
            nn.Sequential(
                nn.Linear(state_dim, router_hidden),
                nn.ReLU(),
                nn.Linear(router_hidden, router_hidden),
                nn.ReLU(),
                nn.Linear(router_hidden, self.n_mlps),
            )
            for _ in range(state_dim)
        ])

        # Learnable coefficients for each selected MLP
        # (the router picks WHICH MLP, the coefficient scales HOW MUCH)
        self.coefficients = nn.ParameterList([
            nn.Parameter(torch.zeros(self.n_mlps))
            for _ in range(state_dim)
        ])

    def forward(self, X, temperature=1.0, hard=True):
        """
        Forward pass: route state data through selected MLPs.

        Args:
            X: state tensor (batch, state_dim)
            temperature: Gumbel-Softmax temperature
            hard: use Straight-Through Estimator

        Returns:
            dXdt_pred: predicted derivatives (batch, state_dim)
            gates: routing decisions (state_dim, batch, n_mlps)
        """
        batch_size = X.shape[0]

        # Evaluate ALL MLPs on the state data
        # (we need all outputs to compute the weighted sum)
        mlp_outputs = []  # will be (n_mlps, batch, 1)
        for i, (model, cols) in enumerate(zip(self.mlp_models, self.input_columns)):
            x_in = X[:, cols]  # select input columns
            out = model(x_in)  # (batch, 1)
            mlp_outputs.append(out)

        # Stack: (batch, n_mlps)
        mlp_out_matrix = torch.cat(mlp_outputs, dim=1)

        # For each state derivative, compute routing + output
        dXdt_pred = []
        all_gates = []

        for d in range(self.state_dim):
            # Router produces logits over MLPs
            logits = self.routers[d](X)  # (batch, n_mlps)

            # Gumbel-Softmax: differentiable discrete selection
            gate = gumbel_softmax(logits, temperature=temperature, hard=hard)
            all_gates.append(gate)

            # Weighted sum: gate selects MLP, coefficients scale it
            # gate: (batch, n_mlps), coefficients: (n_mlps,)
            # mlp_out_matrix: (batch, n_mlps)
            coeffs = self.coefficients[d]  # (n_mlps,)
            weighted = (gate * coeffs.unsqueeze(0)) * mlp_out_matrix  # (batch, n_mlps)
            dxdt_d = weighted.sum(dim=1, keepdim=True)  # (batch, 1)
            dXdt_pred.append(dxdt_d)

        dXdt_pred = torch.cat(dXdt_pred, dim=1)  # (batch, state_dim)
        all_gates = torch.stack(all_gates, dim=0)  # (state_dim, batch, n_mlps)

        return dXdt_pred, all_gates

    def get_routing_summary(self, X, temperature=0.01):
        """Get a summary of which MLPs are selected for which derivatives."""
        self.eval()
        with torch.no_grad():
            _, gates = self.forward(X, temperature=temperature, hard=True)

        # Average gate activations across batch
        avg_gates = gates.mean(dim=1)  # (state_dim, n_mlps)

        summary = {}
        state_labels = ["dx/dt", "dv/dt"]
        for d in range(self.state_dim):
            selected = []
            for i in range(self.n_mlps):
                activation = avg_gates[d, i].item()
                coeff = self.coefficients[d][i].item()
                if activation > 0.01:  # at least 1% selection
                    selected.append({
                        "name": self.mlp_names[i],
                        "activation": activation,
                        "coefficient": coeff,
                    })
            # Sort by activation
            selected.sort(key=lambda x: -x["activation"])
            summary[state_labels[d]] = selected

        return summary


# ─── Training ────────────────────────────────────────────────────────────────

def train_router(
    router, X_train, dXdt_train, X_val, dXdt_val,
    epochs=2000, lr=3e-3,
    tau_start=5.0, tau_end=0.1, tau_anneal_epochs=1500,
    log_every=100, device="cpu",
):
    """
    Train the Gumbel-Softmax router end-to-end.

    Temperature annealing schedule:
        - Start with high τ (soft selection, easy gradients)
        - Linearly anneal to low τ (hard selection, discrete choices)
    """
    router = router.to(device)
    X_train = torch.from_numpy(X_train.astype(np.float32)).to(device)
    dXdt_train = torch.from_numpy(dXdt_train.astype(np.float32)).to(device)
    X_val = torch.from_numpy(X_val.astype(np.float32)).to(device)
    dXdt_val = torch.from_numpy(dXdt_val.astype(np.float32)).to(device)

    # Only optimize router parameters + coefficients (MLPs are frozen)
    optimizer = optim.Adam(
        list(router.routers.parameters()) + list(router.coefficients.parameters()),
        lr=lr,
    )
    scheduler = optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=epochs)
    criterion = nn.MSELoss()

    history = {"train_loss": [], "val_loss": [], "temperature": [], "epoch": []}

    for epoch in range(1, epochs + 1):
        # Temperature annealing: linear decay
        if epoch <= tau_anneal_epochs:
            tau = tau_start - (tau_start - tau_end) * (epoch / tau_anneal_epochs)
        else:
            tau = tau_end

        # ── Train ──
        router.train()
        pred, gates = router(X_train, temperature=tau, hard=True)
        loss = criterion(pred, dXdt_train)

        # Optional: add entropy regularization to encourage exploration early on
        if tau > 1.0:
            gate_probs = gates.mean(dim=1)  # (state_dim, n_mlps)
            entropy = -(gate_probs * (gate_probs + 1e-10).log()).sum()
            loss = loss - 0.01 * entropy  # encourage exploration

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        scheduler.step()

        # ── Log ──
        if epoch % log_every == 0 or epoch == 1:
            router.eval()
            with torch.no_grad():
                val_pred, _ = router(X_val, temperature=tau, hard=True)
                val_loss = criterion(val_pred, dXdt_val)

            history["train_loss"].append(loss.item())
            history["val_loss"].append(val_loss.item())
            history["temperature"].append(tau)
            history["epoch"].append(epoch)

            print(
                f"  Epoch {epoch:>5d} | τ={tau:.3f} | "
                f"Train: {loss.item():.6f} | Val: {val_loss.item():.6f}"
            )

    return history


# ─── Build Router from Checkpoints ──────────────────────────────────────────

def build_router_from_checkpoints(repo_id="pandevim/cs810", state_dim=2):
    mlp_modules = []
    mlp_configs = []
    unary_names = ["identity", "cos", "sin"]
    binary_names = ["add", "mul"]

    for name in unary_names:
        try:
            path = hf_hub_download(repo_id=repo_id,
                                   filename=f"phase1/checkpoints/mlp_{name}.pt")
        except Exception:
            print(f"  ⚠ Checkpoint not found on HF: mlp_{name}.pt, skipping")
            continue
        model, input_dim = load_grokked_mlp(path)
        mlp_modules.append((f"{name}(x)", model, [0]))
        mlp_configs.append({"input_dim": input_dim})
        model2, _ = load_grokked_mlp(path)
        mlp_modules.append((f"{name}(v)", model2, [1]))
        mlp_configs.append({"input_dim": input_dim})

    for name in binary_names:
        try:
            path = hf_hub_download(repo_id=repo_id,
                                   filename=f"phase1/checkpoints/mlp_{name}.pt")
        except Exception:
            print(f"  ⚠ Checkpoint not found on HF: mlp_{name}.pt, skipping")
            continue
        model, input_dim = load_grokked_mlp(path)
        mlp_modules.append((f"{name}(x,v)", model, [0, 1]))
        mlp_configs.append({"input_dim": input_dim})

    return GumbelRouter(state_dim=state_dim, mlp_modules=mlp_modules,
                        mlp_configs=mlp_configs, router_hidden=64)

In [17]:
# ── Load ODE data ──
print("Loading oscillator data...")
path = hf_hub_download(repo_id=HF_REPO_ID, filename="phase1/data/oscillator_data.npz")
data = np.load(path)
t = data["t"]
X = data["X"]
dXdt = data["dXdt"]
k_true = float(data["k"])
c_true = float(data["c"])
print(f"  True params: k={k_true}, c={c_true}")
print(f"  Data shape: X={X.shape}, dXdt={dXdt.shape}\n")

# Train/val split (80/20)
n = len(t)
n_train = int(0.8 * n)
idx = np.random.default_rng(42).permutation(n)
train_idx, val_idx = idx[:n_train], idx[n_train:]

X_train, X_val = X[train_idx], X[val_idx]
dXdt_train, dXdt_val = dXdt[train_idx], dXdt[val_idx]

# ── Build router ──
print("Building Gumbel-Softmax router from grokked MLPs...")
router = build_router_from_checkpoints(HF_REPO_ID, state_dim=2)

print(f"  Library size: {router.n_mlps} MLPs")
print(f"  MLP names: {router.mlp_names}")
print(f"  Router params: {sum(p.numel() for p in router.parameters() if p.requires_grad)}")
print(f"  Frozen MLP params: {sum(p.numel() for p in router.parameters() if not p.requires_grad)}\n")

# ── Train ──
print("=" * 60)
print("  TRAINING GUMBEL-SOFTMAX ROUTER")
print("=" * 60)

history = train_router(
    router, X_train, dXdt_train, X_val, dXdt_val,
    epochs=3000,
    lr=3e-3,
    tau_start=5.0,
    tau_end=0.05,
    tau_anneal_epochs=2000,
    log_every=200,
    device=device,
)

# ── Results ──
print("\n" + "=" * 60)
print("  ROUTING RESULTS")
print("=" * 60)

# Move to CPU for analysis
router = router.cpu()
X_tensor = torch.from_numpy(X.astype(np.float32))

summary = router.get_routing_summary(X_tensor, temperature=0.01)

for deriv_name, selected in summary.items():
    print(f"\n  {deriv_name}:")
    for s in selected:
        print(f"    {s['name']:>15s}: "
              f"selected {s['activation']*100:.1f}% of the time, "
              f"coefficient = {s['coefficient']:+.4f}")

# Print as equations
print("\n" + "=" * 60)
print("  DISCOVERED EQUATIONS (Gumbel-Softmax)")
print("=" * 60)

state_labels = ["ẋ = dx/dt", "v̇ = dv/dt"]
for d, label in enumerate(state_labels):
    deriv_key = ["dx/dt", "dv/dt"][d]
    selected = summary[deriv_key]
    terms = []
    for s in selected:
        if abs(s["coefficient"]) > 0.01:
            terms.append(f"{s['coefficient']:+.4f} · {s['name']}")
    eq_str = " ".join(terms) if terms else "0"
    print(f"\n  {label} = {eq_str}")

print(f"\n  TRUE EQUATIONS:")
print(f"  ẋ = +1.0 · v")
print(f"  v̇ = -{k_true:.1f} · x  -{c_true:.1f} · v")

# ── Save router checkpoint ──
buf = io.BytesIO()
torch.save({"router_state_dict": router.state_dict(),
            "mlp_names": router.mlp_names,
            "state_dim": 2}, buf)
api.upload_file(path_or_fileobj=io.BytesIO(buf.getvalue()),
                path_in_repo="phase1/checkpoints/router.pt",
                repo_id=HF_REPO_ID, commit_message="Add Gumbel-Softmax router checkpoint")

# ── Save router results + history ──
buf = io.BytesIO()
np.savez(buf,
         mlp_names=np.array(router.mlp_names),
         summary=np.array(summary, dtype=object),
         history=np.array(history, dtype=object))
api.upload_file(path_or_fileobj=io.BytesIO(buf.getvalue()),
                path_in_repo="phase1/data/router_results.npz",
                repo_id=HF_REPO_ID, commit_message="Add Gumbel-Softmax router results")

# ── Plot training curves ──
fig, axes = plt.subplots(1, 3, figsize=(16, 5))

# Loss curves
axes[0].semilogy(history["epoch"], history["train_loss"], "b-", alpha=0.8, label="Train")
axes[0].semilogy(history["epoch"], history["val_loss"], "r-", alpha=0.8, label="Val")
axes[0].set_xlabel("Epoch")
axes[0].set_ylabel("MSE Loss (log)")
axes[0].set_title("Router Training Loss", fontweight="bold")
axes[0].legend()
axes[0].grid(True, alpha=0.3)

# Temperature schedule
axes[1].plot(history["epoch"], history["temperature"], "g-", linewidth=2)
axes[1].set_xlabel("Epoch")
axes[1].set_ylabel("Temperature τ")
axes[1].set_title("Temperature Annealing", fontweight="bold")
axes[1].grid(True, alpha=0.3)

# Final routing bar chart
ax = axes[2]
n_mlps = router.n_mlps
x_pos = np.arange(n_mlps)
width = 0.35

for d, (deriv_key, color, label) in enumerate(zip(
    ["dx/dt", "dv/dt"], ["#4CAF50", "#F44336"], ["dx/dt", "dv/dt"]
)):
    activations = np.zeros(n_mlps)
    for s in summary[deriv_key]:
        idx = router.mlp_names.index(s["name"])
        activations[idx] = s["activation"] * s["coefficient"]
    ax.bar(x_pos + d * width, activations, width, color=color,
            edgecolor="black", linewidth=0.5, label=label, alpha=0.8)

ax.set_xticks(x_pos + width / 2)
ax.set_xticklabels(router.mlp_names, rotation=45, ha="right", fontsize=8)
ax.set_ylabel("Activation × Coefficient")
ax.set_title("Router Selections", fontweight="bold")
ax.axhline(y=0, color="black", linewidth=0.5)
ax.legend()
ax.grid(True, alpha=0.3, axis="y")

plt.suptitle("Gumbel-Softmax Router — Training & Results",
              fontsize=14, fontweight="bold")
plt.tight_layout()
buf = io.BytesIO()
plt.savefig(buf, format="png", dpi=150, bbox_inches="tight")
plt.close()
api.upload_file(path_or_fileobj=io.BytesIO(buf.getvalue()),
                path_in_repo="phase1/plots/router_results.png",
                repo_id=HF_REPO_ID, commit_message="Add router results plot")
print("\n  ✓ Router results plot uploaded to HuggingFace")

# ── Compare with STLSQ if available ──
try:
    stlsq_path = hf_hub_download(repo_id=HF_REPO_ID,
                                 filename="phase1/data/discovery_results.npz")
except Exception:
    stlsq_path = None
if stlsq_path is not None:
    print("\n" + "=" * 60)
    print("  COMPARISON: STLSQ vs GUMBEL-SOFTMAX")
    print("=" * 60)
    stlsq_results = np.load(stlsq_path, allow_pickle=True)
    stlsq_xi = stlsq_results["xi"]
    stlsq_names = list(stlsq_results["term_names"])

    print("\n  STLSQ discovered:")
    for col, label in enumerate(state_labels):
        active = np.abs(stlsq_xi[:, col]) > 1e-10
        terms = [f"{stlsq_xi[i, col]:+.4f}·{stlsq_names[i]}"
                  for i in range(len(stlsq_names)) if active[i]]
        print(f"    {label} = {' '.join(terms) if terms else '0'}")

    print("\n  Gumbel-Softmax discovered:")
    for d, label in enumerate(state_labels):
        deriv_key = ["dx/dt", "dv/dt"][d]
        terms = [f"{s['coefficient']:+.4f}·{s['name']}"
                  for s in summary[deriv_key] if abs(s["coefficient"]) > 0.01]
        print(f"    {label} = {' '.join(terms) if terms else '0'}")

    print(f"\n  TRUE: ẋ = +1.0·v, v̇ = -{k_true:.1f}·x -{c_true:.1f}·v")

print("\n✓ Gumbel-Softmax experiment complete!")

Loading oscillator data...


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:93: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


phase1/data/oscillator_data.npz:   0%|          | 0.00/25.2k [00:00<?, ?B/s]

  True params: k=1.0, c=0.1
  Data shape: X=(600, 2), dXdt=(600, 2)

Building Gumbel-Softmax router from grokked MLPs...


phase1/checkpoints/mlp_identity.pt:   0%|          | 0.00/70.5k [00:00<?, ?B/s]

phase1/checkpoints/mlp_cos.pt:   0%|          | 0.00/70.5k [00:00<?, ?B/s]

phase1/checkpoints/mlp_sin.pt:   0%|          | 0.00/70.5k [00:00<?, ?B/s]

phase1/checkpoints/mlp_add.pt:   0%|          | 0.00/71.0k [00:00<?, ?B/s]

phase1/checkpoints/mlp_mul.pt:   0%|          | 0.00/71.0k [00:00<?, ?B/s]

  Library size: 8 MLPs
  MLP names: ['identity(x)', 'identity(v)', 'cos(x)', 'cos(v)', 'sin(x)', 'sin(v)', 'add(x,v)', 'mul(x,v)']
  Router params: 9760
  Frozen MLP params: 135432

  TRAINING GUMBEL-SOFTMAX ROUTER
  Epoch     1 | τ=4.998 | Train: 0.116902 | Val: 0.160407
  Epoch   200 | τ=4.505 | Train: -0.015704 | Val: 0.023127
  Epoch   400 | τ=4.010 | Train: -0.026345 | Val: 0.014605
  Epoch   600 | τ=3.515 | Train: -0.027749 | Val: 0.012353
  Epoch   800 | τ=3.020 | Train: -0.030309 | Val: 0.011166
  Epoch  1000 | τ=2.525 | Train: -0.032358 | Val: 0.009696
  Epoch  1200 | τ=2.030 | Train: -0.032881 | Val: 0.008721
  Epoch  1400 | τ=1.535 | Train: -0.033168 | Val: 0.008791
  Epoch  1600 | τ=1.040 | Train: -0.033919 | Val: 0.007489
  Epoch  1800 | τ=0.545 | Train: 0.006027 | Val: 0.007220
  Epoch  2000 | τ=0.050 | Train: 0.004940 | Val: 0.005666
  Epoch  2200 | τ=0.050 | Train: 0.004710 | Val: 0.005369
  Epoch  2400 | τ=0.050 | Train: 0.004624 | Val: 0.005504
  Epoch  2600 | τ=0.050

Uploading files as a binary IO buffer is not supported by Xet Storage. Falling back to HTTP upload.
Uploading files as a binary IO buffer is not supported by Xet Storage. Falling back to HTTP upload.
Uploading files as a binary IO buffer is not supported by Xet Storage. Falling back to HTTP upload.



  ✓ Router results plot uploaded to HuggingFace


phase1/data/discovery_results.npz:   0%|          | 0.00/1.50k [00:00<?, ?B/s]


  COMPARISON: STLSQ vs GUMBEL-SOFTMAX

  STLSQ discovered:
    ẋ = dx/dt = -0.3325·identity(x) +0.6675·identity(v) +0.3324·add(x,v)
    v̇ = dv/dt = -0.6322·identity(x) +0.2677·identity(v) -0.3676·add(x,v)

  Gumbel-Softmax discovered:
    ẋ = dx/dt = +1.0000·identity(v) -0.3764·cos(v) -0.7670·cos(x) +0.2997·add(x,v) -0.7885·mul(x,v) -0.4449·identity(x) -0.1939·sin(x)
    v̇ = dv/dt = -0.9940·identity(x) +0.3412·cos(x) -0.4907·add(x,v) -0.8817·sin(x) +0.2588·identity(v) -0.3183·sin(v) +0.7209·cos(v) +0.0448·mul(x,v)

  TRUE: ẋ = +1.0·v, v̇ = -1.0·x -0.1·v

✓ Gumbel-Softmax experiment complete!


# Analysis from Phase 5: the router did not commit
The headline failure is visible in the bar chart and the routing percentages: no MLP is selected more than 43.8% of the time for any derivative. A working Gumbel-Softmax run should end with near-one-hot activations (>95% on one term per derivative). Yours looks like a softmax soup.

## What actually worked
- dx/dt → identity(v) with coefficient +1.0000 at 43.8% activation. That coefficient is exact. The router has found the right term, it just hasn't fully committed.
- dv/dt → identity(x) with coefficient -0.9940 at 19.2% activation. Also structurally correct (true value -1.0).

## What broke
1. The damping term is gone. True v̇ has -0.1·v. Your discovered v̇ has +0.2588·identity(v) at 9.7% — wrong sign, wrong magnitude. Damping is 10x smaller than the restoring force, and without a sparsity prior it gets drowned by noise spread across 7 other terms.

2. The soft-mixture escape hatch. With high τ, the router computes a weighted average over all MLPs and learns per-MLP coefficients that combine to fit the data. Nothing in the loss pushes the logits apart. At τ=0.05 you sample discretely but the underlying distribution never concentrated, so get_routing_summary at τ=0.01 still shows 8 terms with non-trivial mass.

3. The collinearity problem is still here — just wearing a new costume. STLSQ split coefficients between identity(x/v) and add(x,v). The router is doing the same thing probabilistically: add(x,v) is selected 10.2% / 14.2% of the time with meaningful coefficients. The earlier claim that "Gumbel-Softmax sidesteps rank deficiency" was mechanically correct but incomplete — rank deficiency became probability-mass splitting.

4. The discontinuity at epoch ~1800 in the loss plot. Train loss jumps from ~negative values (which itself is suspicious — likely a logging artifact of soft-selection masking) to a higher plateau the moment τ hits its floor. That jump is the moment the soft mixture stops hiding the lack of commitment. A well-trained router should show no such discontinuity — it means the soft and hard regimes disagree.

5. Final MSE (~5e-3) is ~1000x worse than STLSQ's 1e-6. Even though STLSQ's coefficients were "spread," its predictions were essentially exact. The router's predictions are genuinely worse, not just less interpretable.

## Verdict
- Experiment 1 (STLSQ): correct answer, ugly presentation.
- Experiment 3 (this run): ugly presentation and wrong answer (missing damping, wrong sign on identity(v) in v̇).

> The router has the capacity to find the right answer — it nailed +1.0·v — but the training objective lets it settle for a soft mixture.

## What I'd try next, in priority order
1. Entropy regularization on the routing logits. Add λ · H(π) with negative sign to the loss (i.e., penalize high entropy). Without this, nothing pushes the router toward one-hot. This alone is likely to fix most of what you see.
2. Longer soak at low τ. Currently τ anneals 5→0.05 over epochs 0-2000, then stays at 0.05 for 1000 epochs. Try extending the low-τ phase to 3000+ epochs, or going to τ=0.01.
3. Drop add(x,v) from the library (the one-line fix from Experiment 2). This removes the probability-mass-splitting attractor and gives the router a cleaner landscape.
4. Sanity check the coefficient parametrization. If the per-MLP coefficient is learned independently of routing probability, the router can minimize loss by picking large coefficients on rarely-selected MLPs. Coupling them (e.g., expected-value loss) may help.
5. Only then worry about the damping term — steps 1-3 may recover it automatically once the router commits.

**Initial Conclusion:** The core issue is missing sparsity pressure, not the method. Gumbel-Softmax without entropy regularization is just a fancy softmax.


# Phase 5: The Gumbel-Softmax Router experiment — Experiment 2

Second pass at the end-to-end differentiable router after Experiment 1 failed
to commit (no MLP selected more than ~44% of the time, damping term lost).

### What changed since Experiment 1

1. **State-independent routing.** The router is no longer a state-dependent
   MLP (`Linear → ReLU → Linear → ReLU → Linear`). It is now a single
   learnable logit vector per derivative, broadcast across the batch.
   - Matches the SINDy assumption that a single term governs the dynamics
     everywhere in state space, not region-by-region.
   - Param count drops from ~9,760 → 16 (2 derivatives × 8 MLPs).
   - Each batch sample still gets its own Gumbel noise draw, so exploration
     is preserved.
2. **Commitment entropy penalty (correct sign, scheduled weight).** The
   previous run *subtracted* entropy from the loss (pushing distributions
   toward *higher* entropy — the wrong direction). Now we *add* a
   temperature-scheduled entropy penalty:
entropy_weight = 0.05 · max(0, 1 − τ / τ_start)
loss += entropy_weight · H(softmax(logits))


- At τ = 5.0 (start): weight = 0 → pure exploration via Gumbel noise.
- At τ = 0.05 (end): weight ≈ 0.0495 → strong pressure toward one-hot.

**What to look for (vs. Experiment 1):**

- `identity(v)` activation on `dx/dt` should jump from 43.8% → ~100%.
- `identity(x)` activation on `dv/dt` should similarly concentrate toward 100%.
- Loss curve should be smooth through the τ-anneal handoff (no jump at
epoch ~1800 like last time).
- Open question: the damping term (`-0.1·v` in `v̇`) is ~10× smaller than
the restoring force. One-hot commitment per derivative may force the
router to miss it. If it does, next step is a top-k variant that allows
2 active terms per derivative.

In [23]:
"""
Gumbel-Softmax Neural Router for MLP Library Selection.

Instead of using STLSQ sparse regression (a post-hoc, non-gradient method),
we train a "router" network that learns to select which grokked MLP
to apply at each state, using Gumbel-Softmax for differentiable discrete choices.

Architecture:
    State [x, v] → Router Network → Gumbel-Softmax → Gate (one-hot) → Σ(gate_i · MLP_i(state))

The router learns WHICH MLP explains each component of the dynamics,
while the grokked MLPs provide the actual function approximations.

Key technique: Straight-Through Estimator (STE)
    Forward pass: hard one-hot selection (only 1 MLP computes)
    Backward pass: smooth Gumbel-Softmax gradients flow through all MLPs
"""

# ─── Gumbel-Softmax ──────────────────────────────────────────────────────────

def gumbel_softmax(logits, temperature=1.0, hard=True):
    """
    Sample from the Gumbel-Softmax distribution.

    Args:
        logits: (batch, n_choices) unnormalized log-probabilities
        temperature: τ — controls sharpness. High=soft, Low=hard.
        hard: if True, use Straight-Through Estimator (STE)
              Forward: hard one-hot. Backward: soft gradients.

    Returns:
        (batch, n_choices) — one-hot if hard=True, soft probabilities otherwise
    """
    # Sample Gumbel noise: g_i = -log(-log(u_i)), u_i ~ Uniform(0,1)
    gumbel_noise = -torch.log(-torch.log(
        torch.rand_like(logits).clamp(min=1e-10)
    ))

    # Add noise to logits and apply temperature-scaled softmax
    y_soft = F.softmax((logits + gumbel_noise) / temperature, dim=-1)

    if hard:
        # Straight-Through Estimator
        # Forward: hard one-hot (argmax)
        index = y_soft.max(dim=-1, keepdim=True)[1]
        y_hard = torch.zeros_like(logits).scatter_(-1, index, 1.0)
        # Backward: pretend we used y_soft (gradient flows through)
        return y_hard - y_soft.detach() + y_soft
    else:
        return y_soft


# ─── Router Network ──────────────────────────────────────────────────────────

class GumbelRouter(nn.Module):
    """
    A router that learns to select which grokked MLP to apply for each
    component of the dynamics.

    For a system with state_dim=2 (x, v), we learn TWO routing decisions:
        - Which MLP combination explains dx/dt?
        - Which MLP combination explains dv/dt?

    Routing is STATE-INDEPENDENT: one logit vector per derivative, shared
    across the whole batch. This matches the SINDy assumption that a single
    term governs the dynamics everywhere in state space.
    """

    def __init__(self, state_dim, mlp_modules, mlp_configs, router_hidden=64):
        """
        Args:
            state_dim: dimension of the state space (e.g., 2 for [x, v])
            mlp_modules: list of (name, model, input_columns) tuples
            mlp_configs: list of dicts with 'input_dim' for each MLP
            router_hidden: unused (kept for backward-compat with old checkpoints)
        """
        super().__init__()
        self.state_dim = state_dim
        self.n_mlps = len(mlp_modules)
        self.mlp_names = [m[0] for m in mlp_modules]
        self.mlp_models = nn.ModuleList([m[1] for m in mlp_modules])
        self.input_columns = [m[2] for m in mlp_modules]

        # Freeze grokked MLPs — we don't want to un-grok them!
        for model in self.mlp_models:
            for param in model.parameters():
                param.requires_grad = False

        # One logit vector per state derivative (state-independent router).
        # Same decision applies across the whole batch — no state-dependent MLP.
        self.routers = nn.ParameterList([
            nn.Parameter(torch.zeros(self.n_mlps))
            for _ in range(state_dim)
        ])

        # Learnable coefficients for each selected MLP
        # (the router picks WHICH MLP, the coefficient scales HOW MUCH)
        self.coefficients = nn.ParameterList([
            nn.Parameter(torch.zeros(self.n_mlps))
            for _ in range(state_dim)
        ])

    def forward(self, X, temperature=1.0, hard=True):
        """
        Forward pass: route state data through selected MLPs.

        Args:
            X: state tensor (batch, state_dim)
            temperature: Gumbel-Softmax temperature
            hard: use Straight-Through Estimator

        Returns:
            dXdt_pred: predicted derivatives (batch, state_dim)
            gates: routing decisions (state_dim, batch, n_mlps)
        """
        batch_size = X.shape[0]

        # Evaluate ALL MLPs on the state data
        # (we need all outputs to compute the weighted sum)
        mlp_outputs = []  # will be (n_mlps, batch, 1)
        for i, (model, cols) in enumerate(zip(self.mlp_models, self.input_columns)):
            x_in = X[:, cols]  # select input columns
            out = model(x_in)  # (batch, 1)
            mlp_outputs.append(out)

        # Stack: (batch, n_mlps)
        mlp_out_matrix = torch.cat(mlp_outputs, dim=1)

        # For each state derivative, compute routing + output
        dXdt_pred = []
        all_gates = []

        for d in range(self.state_dim):
            # State-independent logits, broadcast to batch so each sample
            # still gets an independent Gumbel sample (exploration).
            logits = self.routers[d].unsqueeze(0).expand(batch_size, -1)  # (batch, n_mlps)

            # Gumbel-Softmax: differentiable discrete selection
            gate = gumbel_softmax(logits, temperature=temperature, hard=hard)
            all_gates.append(gate)

            # Weighted sum: gate selects MLP, coefficients scale it
            # gate: (batch, n_mlps), coefficients: (n_mlps,)
            # mlp_out_matrix: (batch, n_mlps)
            coeffs = self.coefficients[d]  # (n_mlps,)
            weighted = (gate * coeffs.unsqueeze(0)) * mlp_out_matrix  # (batch, n_mlps)
            dxdt_d = weighted.sum(dim=1, keepdim=True)  # (batch, 1)
            dXdt_pred.append(dxdt_d)

        dXdt_pred = torch.cat(dXdt_pred, dim=1)  # (batch, state_dim)
        all_gates = torch.stack(all_gates, dim=0)  # (state_dim, batch, n_mlps)

        return dXdt_pred, all_gates

    def get_routing_summary(self, X, temperature=0.01):
        """Get a summary of which MLPs are selected for which derivatives."""
        self.eval()
        with torch.no_grad():
            _, gates = self.forward(X, temperature=temperature, hard=True)

        # Average gate activations across batch
        avg_gates = gates.mean(dim=1)  # (state_dim, n_mlps)

        summary = {}
        state_labels = ["dx/dt", "dv/dt"]
        for d in range(self.state_dim):
            selected = []
            for i in range(self.n_mlps):
                activation = avg_gates[d, i].item()
                coeff = self.coefficients[d][i].item()
                if activation > 0.01:  # at least 1% selection
                    selected.append({
                        "name": self.mlp_names[i],
                        "activation": activation,
                        "coefficient": coeff,
                    })
            # Sort by activation
            selected.sort(key=lambda x: -x["activation"])
            summary[state_labels[d]] = selected

        return summary


# ─── Training ────────────────────────────────────────────────────────────────

def train_router(
    router, X_train, dXdt_train, X_val, dXdt_val,
    epochs=2000, lr=3e-3,
    tau_start=5.0, tau_end=0.1, tau_anneal_epochs=1500,
    log_every=100, device="cpu",
):
    """
    Train the Gumbel-Softmax router end-to-end.

    Temperature annealing schedule:
        - Start with high τ (soft selection, easy gradients)
        - Linearly anneal to low τ (hard selection, discrete choices)
    """
    router = router.to(device)
    X_train = torch.from_numpy(X_train.astype(np.float32)).to(device)
    dXdt_train = torch.from_numpy(dXdt_train.astype(np.float32)).to(device)
    X_val = torch.from_numpy(X_val.astype(np.float32)).to(device)
    dXdt_val = torch.from_numpy(dXdt_val.astype(np.float32)).to(device)

    # Only optimize router parameters + coefficients (MLPs are frozen)
    optimizer = optim.Adam(
        list(router.routers.parameters()) + list(router.coefficients.parameters()),
        lr=lr,
    )
    scheduler = optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=epochs)
    criterion = nn.MSELoss()

    history = {"train_loss": [], "val_loss": [], "temperature": [], "epoch": []}

    for epoch in range(1, epochs + 1):
        # Temperature annealing: linear decay
        if epoch <= tau_anneal_epochs:
            tau = tau_start - (tau_start - tau_end) * (epoch / tau_anneal_epochs)
        else:
            tau = tau_end

        # ── Train ──
        router.train()
        pred, gates = router(X_train, temperature=tau, hard=True)
        loss = criterion(pred, dXdt_train)

        # Commitment entropy penalty on the routing distributions.
        # Weight ramps up as temperature anneals down: early on we rely on
        # Gumbel noise for exploration, late we force one-hot commitment.
        entropy_weight = 0.05 * max(0.0, 1.0 - tau / tau_start)
        if entropy_weight > 0:
            entropy_loss = 0.0
            for d in range(router.state_dim):
                probs = F.softmax(router.routers[d], dim=-1)
                entropy_loss = entropy_loss + -(probs * (probs + 1e-10).log()).sum()
            loss = loss + entropy_weight * entropy_loss

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        scheduler.step()

        # ── Log ──
        if epoch % log_every == 0 or epoch == 1:
            router.eval()
            with torch.no_grad():
                val_pred, _ = router(X_val, temperature=tau, hard=True)
                val_loss = criterion(val_pred, dXdt_val)

            history["train_loss"].append(loss.item())
            history["val_loss"].append(val_loss.item())
            history["temperature"].append(tau)
            history["epoch"].append(epoch)

            print(
                f"  Epoch {epoch:>5d} | τ={tau:.3f} | "
                f"Train: {loss.item():.6f} | Val: {val_loss.item():.6f}"
            )

    return history


# ─── Build Router from Checkpoints ──────────────────────────────────────────

def build_router_from_checkpoints(repo_id="pandevim/cs810", state_dim=2):
    mlp_modules = []
    mlp_configs = []
    unary_names = ["identity", "cos", "sin"]
    binary_names = ["add", "mul"]

    for name in unary_names:
        try:
            path = hf_hub_download(repo_id=repo_id,
                                   filename=f"phase1/checkpoints/mlp_{name}.pt")
        except Exception:
            print(f"  ⚠ Checkpoint not found on HF: mlp_{name}.pt, skipping")
            continue
        model, input_dim = load_grokked_mlp(path)
        mlp_modules.append((f"{name}(x)", model, [0]))
        mlp_configs.append({"input_dim": input_dim})
        model2, _ = load_grokked_mlp(path)
        mlp_modules.append((f"{name}(v)", model2, [1]))
        mlp_configs.append({"input_dim": input_dim})

    for name in binary_names:
        try:
            path = hf_hub_download(repo_id=repo_id,
                                   filename=f"phase1/checkpoints/mlp_{name}.pt")
        except Exception:
            print(f"  ⚠ Checkpoint not found on HF: mlp_{name}.pt, skipping")
            continue
        model, input_dim = load_grokked_mlp(path)
        mlp_modules.append((f"{name}(x,v)", model, [0, 1]))
        mlp_configs.append({"input_dim": input_dim})

    return GumbelRouter(state_dim=state_dim, mlp_modules=mlp_modules,
                        mlp_configs=mlp_configs, router_hidden=64)

In [ ]:
    # Routing is STATE-INDEPENDENT: one logit vector per derivative, shared
    # across the whole batch. This matches the SINDy assumption that a single
    # term governs the dynamics everywhere in state space.


# ─── 1. Patch GumbelRouter Methods ────────────────────────────────────────

def v2_init(self, state_dim, mlp_modules, mlp_configs, router_hidden=64):
    """
    V2 State-Independent Gumbel-Softmax Router for MLP Library Selection.

    Args:
        state_dim: dimension of the state space (e.g., 2 for [x, v])
        mlp_modules: list of (name, model, input_columns) tuples
        mlp_configs: list of dicts with 'input_dim' for each MLP
        router_hidden: unused (kept for backward-compat with old checkpoints)
    """
    super().__init__()
    self.state_dim = state_dim
    self.n_mlps = len(mlp_modules)
    self.mlp_names = [m[0] for m in mlp_modules]
    self.mlp_models = nn.ModuleList([m[1] for m in mlp_modules])
    self.input_columns = [m[2] for m in mlp_modules]

    # Freeze grokked MLPs — we don't want to un-grok them!
    for model in self.mlp_models:
        for param in model.parameters():
            param.requires_grad = False

    # One logit vector per state derivative (state-independent router).
    # Same decision applies across the whole batch — no state-dependent MLP.
    self.routers = nn.ParameterList([
        nn.Parameter(torch.zeros(self.n_mlps))
        for _ in range(state_dim)
    ])

    # Learnable coefficients for each selected MLP
    # (the router picks WHICH MLP, the coefficient scales HOW MUCH)
    self.coefficients = nn.ParameterList([
        nn.Parameter(torch.zeros(self.n_mlps))
        for _ in range(state_dim)
    ])

def v2_forward(self, X, temperature=1.0, hard=True):
    """
    V2 State-Independent Forward Pass

    Forward pass: route state data through selected MLPs.

    Args:
        X: state tensor (batch, state_dim)
        temperature: Gumbel-Softmax temperature
        hard: use Straight-Through Estimator

    Returns:
        dXdt_pred: predicted derivatives (batch, state_dim)
        gates: routing decisions (state_dim, batch, n_mlps)
    """
    batch_size = X.shape[0]

    # Evaluate ALL MLPs on the state data
    # (we need all outputs to compute the weighted sum)
    mlp_outputs = []  # will be (n_mlps, batch, 1)
    for i, (model, cols) in enumerate(zip(self.mlp_models, self.input_columns)):
        x_in = X[:, cols]  # select input columns
        out = model(x_in)  # (batch, 1)
        mlp_outputs.append(out)

    # Stack: (batch, n_mlps)
    mlp_out_matrix = torch.cat(mlp_outputs, dim=1)

    # For each state derivative, compute routing + output
    dXdt_pred = []
    all_gates = []

    for d in range(self.state_dim):
        # State-independent logits, broadcast to batch so each sample
        # still gets an independent Gumbel sample (exploration).
        logits = self.routers[d].unsqueeze(0).expand(batch_size, -1)  # (batch, n_mlps)

        # Gumbel-Softmax: differentiable discrete selection
        gate = gumbel_softmax(logits, temperature=temperature, hard=hard)
        all_gates.append(gate)

        # Weighted sum: gate selects MLP, coefficients scale it
        # gate: (batch, n_mlps), coefficients: (n_mlps,)
        # mlp_out_matrix: (batch, n_mlps)
        coeffs = self.coefficients[d]  # (n_mlps,)
        weighted = (gate * coeffs.unsqueeze(0)) * mlp_out_matrix  # (batch, n_mlps)
        dxdt_d = weighted.sum(dim=1, keepdim=True)  # (batch, 1)
        dXdt_pred.append(dxdt_d)

    dXdt_pred = torch.cat(dXdt_pred, dim=1)  # (batch, state_dim)
    all_gates = torch.stack(all_gates, dim=0)  # (state_dim, batch, n_mlps)

    return dXdt_pred, all_gates

# Apply monkey patches to the class
GumbelRouter.__init__ = v2_init
GumbelRouter.forward = v2_forward


# ─── 2. Patch Training Function ───────────────────────────────────────────

def train_router_v2(
    router, X_train, dXdt_train, X_val, dXdt_val,
    epochs=2000, lr=3e-3,
    tau_start=5.0, tau_end=0.1, tau_anneal_epochs=1500,
    log_every=100, device="cpu",
):
    """
    V2 Training Loop with Commitment Entropy Penalty

    Train the Gumbel-Softmax router end-to-end.

    Temperature annealing schedule:
        - Start with high τ (soft selection, easy gradients)
        - Linearly anneal to low τ (hard selection, discrete choices)
    """
    router = router.to(device)
    X_train = torch.from_numpy(X_train.astype(np.float32)).to(device)
    dXdt_train = torch.from_numpy(dXdt_train.astype(np.float32)).to(device)
    X_val = torch.from_numpy(X_val.astype(np.float32)).to(device)
    dXdt_val = torch.from_numpy(dXdt_val.astype(np.float32)).to(device)

    # Only optimize router parameters + coefficients (MLPs are frozen)
    optimizer = optim.Adam(
        list(router.routers.parameters()) + list(router.coefficients.parameters()),
        lr=lr,
    )
    scheduler = optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=epochs)
    criterion = nn.MSELoss()

    history = {"train_loss": [], "val_loss": [], "temperature": [], "epoch": []}

    for epoch in range(1, epochs + 1):
        # Temperature annealing: linear decay
        if epoch <= tau_anneal_epochs:
            tau = tau_start - (tau_start - tau_end) * (epoch / tau_anneal_epochs)
        else:
            tau = tau_end

        # ── Train ──
        router.train()
        pred, gates = router(X_train, temperature=tau, hard=True)
        loss = criterion(pred, dXdt_train)

        # Commitment entropy penalty on the routing distributions.
        # Weight ramps up as temperature anneals down: early on we rely on
        # Gumbel noise for exploration, late we force one-hot commitment.
        entropy_weight = 0.05 * max(0.0, 1.0 - tau / tau_start)
        if entropy_weight > 0:
            entropy_loss = 0.0
            for d in range(router.state_dim):
                probs = F.softmax(router.routers[d], dim=-1)
                entropy_loss = entropy_loss + -(probs * (probs + 1e-10).log()).sum()
            loss = loss + entropy_weight * entropy_loss

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        scheduler.step()

        # ── Log ──
        if epoch % log_every == 0 or epoch == 1:
            router.eval()
            with torch.no_grad():
                val_pred, _ = router(X_val, temperature=tau, hard=True)
                val_loss = criterion(val_pred, dXdt_val)

            history["train_loss"].append(loss.item())
            history["val_loss"].append(val_loss.item())
            history["temperature"].append(tau)
            history["epoch"].append(epoch)

            print(
                f"  Epoch {epoch:>5d} | τ={tau:.3f} | "
                f"Train: {loss.item():.6f} | Val: {val_loss.item():.6f}"
            )

    return history

# Override the function in the global namespace if needed
train_router = train_router_v2

```txt
============================================================
  ROUTING RESULTS
============================================================

  dx/dt:
             sin(v): selected 99.7% of the time, coefficient = +1.0265

  dv/dt:
             sin(x): selected 99.2% of the time, coefficient = -1.0138

============================================================
  DISCOVERED EQUATIONS (Gumbel-Softmax)
============================================================

  ẋ = dx/dt = +1.0265 · sin(v)

  v̇ = dv/dt = -1.0138 · sin(x)

  TRUE EQUATIONS:
  ẋ = +1.0 · v
  v̇ = -1.0 · x  -0.1 · v

============================================================
  COMPARISON: STLSQ vs GUMBEL-SOFTMAX
============================================================

  STLSQ discovered:
    ẋ = dx/dt = -0.3325·identity(x) +0.6675·identity(v) +0.3324·add(x,v)
    v̇ = dv/dt = -0.6322·identity(x) +0.2677·identity(v) -0.3676·add(x,v)

  Gumbel-Softmax discovered:
    ẋ = dx/dt = +1.0265·sin(v)
    v̇ = dv/dt = -1.0138·sin(x)

  TRUE: ẋ = +1.0·v, v̇ = -1.0·x -0.1·v
```

## Analysis: the router committed cleanly — but to the wrong term

Experiment 2 fixes the problem Experiment 1 failed on (commitment), cleanly
demonstrates a new problem, and confirms a suspicion about the library.

### What worked

- **One-hot routing achieved.** `sin(v)` selected 99.7% of the time for
  `dx/dt`, `sin(x)` selected 99.2% for `dv/dt`. No mass leaking to other
  terms. The entropy penalty + state-independent router did exactly what
  they were designed to do.
- **Smooth loss curve.** The discontinuity at epoch ~1800 from Experiment 1
  is gone. Train loss descends monotonically through the τ handoff, which
  means the soft and hard regimes now agree — the router's logits were
  concentrated *before* the Gumbel samples became effectively deterministic.
- **Router parameter count dropped from 9,760 → 32** (16 logits + 16
  coefficients). The state-independent router is ~300× smaller and generalizes
  better: val loss ~1.2e-3, ~4× better than Experiment 1's ~5e-3.
- **Coefficients are near the true magnitudes** (+1.0265 and -1.0138 vs.
  true ±1.0). That's the right ballpark for the restoring term.

### What broke

1. **The router chose `sin` over `identity`.** True equations are linear in
   `x` and `v`, so `identity(x)` and `identity(v)` should have won. The
   router picked `sin(x)` and `sin(v)` instead. This is a **library
   degeneracy** — for the amplitudes the oscillator visits (|x|, |v| ≲ 1),
   `sin(z) ≈ z` to within ~15% at the peaks. Both basis functions are
   *approximately* collinear, and the entropy penalty broke the tie
   arbitrarily.

   The coefficient inflation (+1.0265 vs. true +1.0) is the tell: since
   `|sin(z)| < |z|`, the coefficient has to absorb the Taylor-series
   shortfall. Had the router picked identity, the coefficient would have
   landed exactly on 1.0.

2. **Damping is still gone.** True `v̇ = -1.0·x - 0.1·v`. Discovered
   `v̇ = -1.0138·sin(x)`. The `-0.1·v` term is ~10× smaller than the
   restoring force and one-hot commitment *cannot* represent a two-term
   equation. This is a structural limitation of the current router, not a
   training failure.

3. **MSE is ~5000× worse than STLSQ.** Final train MSE ~5.6e-3 vs. STLSQ's
   ~1e-6. The sin-vs-identity approximation error dominates. Gumbel-Softmax
   is now more interpretable than STLSQ (two clean one-hot selections vs.
   six-term spread) but less accurate.

4. **The val-loss spike at τ=0.05 (epoch 2000).** A one-epoch jump from
   ~1.2e-3 to ~3.5e-3 when τ hits its floor. This is the residual of
   Experiment 1's problem in miniature — the router's top-1 choice wasn't
   *quite* stable yet when hard sampling became deterministic. It recovers
   within 200 epochs, so it's benign, but a slower anneal would smooth it.

### The deeper insight

**The library has two near-collinearities:**
- *Exact* collinearity: `add(x,v) = identity(x) + identity(v)` — broke STLSQ.
- *Approximate* collinearity: `sin(z) ≈ identity(z)` at small amplitudes —
  broke the one-hot router.

Experiment 1 (STLSQ) hit the first. Experiment 2 (Gumbel-Softmax) sidesteps
the first but hits the second. Neither method is "failing" — they're both
revealing that **the library design is the real bottleneck.** A library with
redundant terms will always put the burden on the optimizer to disambiguate,
and different optimizers fail in different ways.

### What I'd try next, in priority order

1. **Allow top-k routing per derivative** (k=2). One `sin` term cannot
   represent `v̇ = -1·x - 0.1·v`. A two-term selection can. Implementation:
   pick top-2 logits per derivative instead of argmax, pass both through
   with their learned coefficients. This alone should recover damping.

2. **Prefer simpler basis via a complexity prior.** Add a per-term
   "complexity penalty" in the logits — e.g., `identity` gets a log-prior
   bonus of +α over `sin`/`cos`. Breaks the sin/identity tie in favor of
   the simpler (and correct) term. This is a cheap, honest fix: Occam's
   razor baked into the router.

3. **Test the hypothesis directly.** Re-run Experiment 2 with `sin` and
   `cos` *removed* from the library (leaving only `identity`, `add`, `mul`).
   If the router then picks `identity(v)` / `identity(x)` with coefficients
   converging to ±1.0, that confirms the sin-vs-identity degeneracy is the
   sole reason for the wrong selection.

4. **Only then revisit the damping question.** If 1 and 2 don't recover
   `-0.1·v`, the issue is data (600 points may not be enough signal for a
   term 10× smaller than the dominant one) rather than routing.

### Scorecard vs. Experiment 1

| Metric                       | Exp 1 (MLP router)      | Exp 2 (scalar router)  |
|------------------------------|-------------------------|------------------------|
| Router params                | 9,760                   | 32                     |
| Max activation (dx/dt)       | 43.8%                   | 99.7%                  |
| Max activation (dv/dt)       | 19.2%                   | 99.2%                  |
| Val MSE                      | ~5e-3                   | ~1.2e-3                |
| Loss discontinuity at τ drop | Large                   | Small spike, recovers  |
| Equations interpretable?     | No (7-8 terms each)     | Yes (1 term each)      |
| Equations correct?           | Structurally yes        | Wrong basis, damping gone |

Experiment 2 is not yet the right answer, but it is a **cleaner failure**
than Experiment 1 — and a cleaner failure is a better foundation for the
next iteration. We've gone from "router can't commit" to "router commits
confidently to a near-correct but simpler-is-better answer." That's
progress.